File: Solutions/All/217-Contains-Duplicate/ContainsDuplicate.go
package solutions

func containsDuplicate(nums []int) bool {
	uniqueNums := make(map[int]bool)

	for _, num := range nums {
		if _, ok := uniqueNums[num]; ok {
			return true
		}

		uniqueNums[num] = true
	}

	return false
}

File: Solutions/All/217-Contains-Duplicate/ContainsDuplicate.ts
// O(n) time and space solution
function containsDuplicate(nums: number[]): boolean {
  let uniqueNums = new Set<number>();

  for (let num of nums) {
    if (uniqueNums.has(num)) return true;
    else uniqueNums.add(num);
  }

  return false;
}

// Constant space, but 0(n log n) solution
function containsDuplicateConstantSpace(nums: number[]): boolean {
  nums = nums.sort((a, b) => a - b);

  for (let i = 1; i < nums.length; i++) {
    if (nums[i] === nums[i - 1]) return true;
  }

  return false;
}
